### PR TITLE
simplify connection establishment

### LINKED ISSUE #8688: TDP: simplify connection establishment
Currently, when a TDP connection is established between client (browser) and server (windows desktop service), the client must always send a "client username" message (to indicate what OS-user the Teleport user is attempting to log in as) followed by the "client size" message (to indicate the initial screen size when the desktop session is opened).

The messages must be sent in that order, and the backend will establish the RDP connection to the Windows Desktop only after they are received.

@alex-kovoy suggests a nice clean up where:

1. the client username and initial screen size are embedded in the connection string (the backend then has this information before TDP messages start flowing)
2. the client username TDP message can be removed from the protocol entirely as it's not needed again
3. the client screen size message must remain, but is only sent for resize events and not part of connection establishment

This simplifies connection establishment, as the backend no longer has to make assumptions about what messages are coming and in what order they will be received.

### PR DESCRIPTION
**Pull request by ibeckermayer**
Adds the username and initial screen size to the websocket parameter string and sends over TDP during the connection process.

Compatible with https://github.com/gravitational/webapps/pull/480

closes https://github.com/gravitational/teleport/issues/8688

### CODE REVIEW

File: lib/web/desktop.go [hunk 1]
```diff
@@ -48,6 +51,40 @@ func (h *Handler) handleDesktopAccessWebsocket(
...
+		errStr := "missing initial_height"
+		log.Error(errStr)
+		return nil, trace.BadParameter(errStr)
+	}
+	initialHeight, err := strconv.Atoi(initialHeightStr)
+	if err != nil {
+		errStr := "initial_height could not be converted to an integer"
+		log.Error(errStr)
+		return nil, trace.Wrap(err, errStr)
+	}
```
Comment (alex-kovoy) [inline lib/web/desktop.go:84]:
```go
height, err := strconv.Atoi(q.Get("height"))
if err != nil {		
     return nil, trace.Wrap(err, "height is missing or invalid")
}
```

File: lib/web/desktop.go [hunk 2]
```diff
@@ -84,6 +121,11 @@ func (h *Handler) handleDesktopAccessWebsocket(
 	serviceConTLS := tls.Client(serviceCon, ctx.clt.Config())
 	log.Debug("Connected to windows_desktop_service")
 
+	// Write username, initial width, and initial height to the windows desktop service using TDP
+	tdpConn := tdp.NewConn(serviceConTLS)
+	tdpConn.OutputMessage(tdp.ClientUsername{Username: username})
+	tdpConn.OutputMessage(tdp.ClientScreenSpec{Width: uint32(initialWidth), Height: uint32(initialHeight)})
```
Comment (alex-kovoy) [inline lib/web/desktop.go:127]: Missing error handling? 

```golang
if err := tdpConn.OutputMessage(tdp.ClientUsername{Username: username}); err != nil {
	return nil, trace.Wrap(err)
}
```

File: lib/web/desktop.go [hunk 3]
```diff
@@ -31,9 +32,11 @@ import (
 	"github.com/gravitational/teleport/api/types"
 	"github.com/gravitational/teleport/lib/reversetunnel"
 	"github.com/gravitational/teleport/lib/srv/desktop"
+	"github.com/gravitational/teleport/lib/srv/desktop/tdp"
 	"github.com/gravitational/teleport/lib/utils"
 )
 
+// GET /webapi/sites/:site/desktops/:desktopName/connect?access_token=<bearer_token>&username=<username>&initial_width=<initialWidth>&initial_height=<initialHeight>
```
Comment (alex-kovoy) [inline lib/web/desktop.go:39]: this is minor but just for consistency reasons i would rename 

`handleDesktopAccessWebsocket -> to deskopAccessHanlde` :) so it's aligned with all other endpoints.

File: lib/web/desktop.go [hunk 4]
```diff
@@ -48,6 +51,40 @@ func (h *Handler) handleDesktopAccessWebsocket(
 	log := ctx.log.WithField("desktop-uuid", desktopName)
 	log.Debug("New desktop access websocket connection")
 
+	q := r.URL.Query()
+	username := q.Get("username")
+	if username == "" {
+		errStr := "missing username"
+		log.Error(errStr)
```
[the remark this answers is not in the document]
Comment (alex-kovoy) [reply lib/web/desktop.go:58]: or if you do want to keep the logs:

```golang
func (h *Handler) createDesktopConnectionHandle(...) (interface{}, error) { 
	if err := h.getDesktopConnectionHandle(...); err != nil {
		h.log(err) // <- log once
		return nil, trace.Wrap(err)
	}

	return nil, nil
}

func (h *Handler) createDesktopConnection(){
	// code here
}
```

File: lib/web/desktop.go [hunk 5]
```diff
@@ -48,6 +51,40 @@ func (h *Handler) handleDesktopAccessWebsocket(
 	log := ctx.log.WithField("desktop-uuid", desktopName)
```
Comment (zmb3) [inline lib/web/desktop.go:51]:
```suggestion
	log := ctx.log.WithField("desktop-name", desktopName)
```

File: lib/web/desktop.go [hunk 6]
```diff
@@ -48,6 +51,40 @@ func (h *Handler) handleDesktopAccessWebsocket(
...
 	log.Debug("New desktop access websocket connection")
 
+	q := r.URL.Query()
+	username := q.Get("username")
+	if username == "" {
+		errStr := "missing username"
+		log.Error(errStr)
+		return nil, trace.BadParameter(errStr)
+	}
+	initialWidthStr := q.Get("initial_width")
```
[the remark this answers is not in the document]
Comment (zmb3) [reply lib/web/desktop.go:61]: +1

File: lib/web/desktop.go [hunk 7]
```diff
@@ -48,6 +51,40 @@ func (h *Handler) handleDesktopAccessWebsocket(
...
+	if initialWidthStr == "" {
+		errStr := "missing initial_width"
+		log.Error(errStr)
+		return nil, trace.BadParameter(errStr)
+	}
+	initialWidth, err := strconv.Atoi(initialWidthStr)
+	if err != nil {
+		errStr := "initial_width could not be converted to an integer"
+		log.Error(errStr)
+		return nil, trace.Wrap(err, errStr)
```
Comment (zmb3) [inline lib/web/desktop.go:71]: I would use `trace.BadParameter` here.

File: lib/web/desktop.go [hunk 8]
```diff
@@ -84,6 +121,11 @@ func (h *Handler) handleDesktopAccessWebsocket(
 	serviceConTLS := tls.Client(serviceCon, ctx.clt.Config())
 	log.Debug("Connected to windows_desktop_service")
 
+	// Write username, initial width, and initial height to the windows desktop service using TDP
```
Comment (zmb3) [inline lib/web/desktop.go:124]: I don't think this comment is necessary. It's already pretty clear that we're using the `tdp` package and writing a `ClientUsername` and `ClientScreenSpec`.

I would just remove it, then it won't go out of date if we eventually support other (non-Windows) desktop services.

File: lib/web/desktop.go (hunk 3 above)
Comment (zmb3) [reply lib/web/desktop.go:39]: `desktopConnectHandle` maybe?

**Comment by alex-kovoy** [comment]
@ibeckermayer ping

File: lib/web/desktop.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply lib/web/desktop.go:61]: Keeping `username` since its used throughout the frontend and the rust rdp client.

File: lib/web/desktop.go [hunk 9]
```diff
@@ -84,12 +122,22 @@ func (h *Handler) handleDesktopAccessWebsocket(
 	serviceConTLS := tls.Client(serviceCon, ctx.clt.Config())
 	log.Debug("Connected to windows_desktop_service")
 
+	tdpConn := tdp.NewConn(serviceConTLS)
+	err = tdpConn.OutputMessage(tdp.ClientUsername{Username: username})
```
Comment (zmb3) [inline lib/web/desktop.go:126]: Do these messages have to happen in `proxyWebsocketConn` after we set `ws.PayloadType` but before we start proxying data?

Comment (ibeckermayer) [reply lib/web/desktop.go:126]: I don't quite see it... Why might they? AFAIK the `ws.PayloadType` isn't relevant at this point, I'm just writing directly to (4)

Image from iOS

I do wonder, though, if there may be a race condition here: [once I write those two messages](https://github.com/gravitational/teleport/blob/master/lib/srv/desktop/rdp/rdpclient/client.go#L145-L153), the RDP client [connects](https://github.com/gravitational/teleport/blob/d5b3cdd8ebde274fd022015e06538971b1d012b9/lib/srv/desktop/rdp/rdpclient/client.go#L196) and then [begins sending pngs](https://github.com/gravitational/teleport/blob/d5b3cdd8ebde274fd022015e06538971b1d012b9/lib/srv/desktop/rdp/rdpclient/client.go#L231). 

If that process all happens very fast before `proxyWebsocketConn` has run, is there a chance that some of the initial png's are never read and sent over the websocket? The read buffer of the `serviceConTLS` must have some limit...

Comment (zmb3) [reply lib/web/desktop.go:126]: Ah, you're right. Thanks for using my own diagram to teach me :) 

I don't see a race, should be fine.

**Review by zmb3 — APPROVED** [review]

**Review by alex-kovoy — APPROVED** [review]

**Pull request merged**